### PR TITLE
remove scheme from s3 endpoint in config

### LINKED ISSUE #44: S3 endpoint config property should not include the protocol
### Bug Description

When relating the coordinator with something implementing the `s3` interface (`ceph-radosgw` from the `reef/edge` channel in my case), the `endpoint` value on the worker's config wrongly includes the protocol (`http(s)`). The rendered config is:

```yaml
s3:
  access_key_id: __THE_KEY__
  bucket_name: mimir
  endpoint: http://10.18.2.87:80
  region: ''
  secret_access_key: __THE_KEY__
```
 And the mimir service doesn't run, logging this plus a traceback:

```
Endpoint url cannot have fully qualified paths.
initialize s3 client
...
```

Manually modifying the `/etc/mimir/mimir-config.yaml` file to remove `http://` allows Mimir to run.

### To Reproduce

I don't have a _minimum_ reproducer, but this is roughly my setup:

1. Deployed microk8s with ceph as storage backend plus the ceph-radosgw to use as Mimir's object store.

```yaml
machines:
  '0':
  '1':
  '2':
applications:
  microk8s:
    charm: microk8s
    channel: 1.28/stable
    series: jammy
    num_units: 3
    to:
    - '0'
    - '1'
    - '2'
  ceph-mon:
    charm: ceph-mon
    channel: quincy/stable
    num_units: 3
    bindings:
      "": oam-space
    to:
    - '0'
    - '1'
    - '2'
  ceph-osd:
    charm: ceph-osd
    channel: quincy/stable
    num_units: 3
    bindings:
      "": oam-space
    to:
    - '0'
    - '1'
    - '2'
    options:
      osd-devices: /dev/vdb
  ceph-csi:
    charm: ceph-csi
    channel: 1.28/stable
    options:
      provisioner-replicas: 1
      namespace: kube-system
  ceph-radosgw:
     charm: ceph-radosgw
     channel: reef/edge
     to: [lxd:0]
  ntp:
    charm: ntp
    channel: stable
relations:
- - microk8s:juju-info
  - ntp:juju-info
- - ceph-osd:mon
  - ceph-mon:osd
- - ceph-csi:ceph-client
  - ceph-mon:client
- - ceph-csi:kubernetes-info
  - microk8s

2. Create a k8s model using the just deployed microk8s as substrate, call it `cos`.
3. Deployed this bundle to it:

```yaml
bundle: kubernetes
applications:
  alertmanager:
    charm: alertmanager-k8s
    channel: latest/edge
    base: ubuntu@20.04/stable
    scale: 1
    storage:
      data: kubernetes,1,2048M
    trust: true
  catalogue:
    charm: catalogue-k8s
    channel: latest/stable
    base: ubuntu@20.04/stable
    scale: 1
    trust: true
  grafana:
    charm: grafana-k8s
    channel: latest/stable
    base: ubuntu@20.04/stable
    resources:
      grafana-image: 67
      litestream-image: 43
    scale: 1
    storage:
      database: kubernetes,1,2048M
    trust: true
  mimir:
    charm: mimir-coordinator-k8s
    channel: edge
    revision: 6
    resources:
      nginx-image: 2
    scale: 1
    constraints: arch=amd64
    trust: true
  mimir-worker:
    charm: mimir-worker-k8s
    channel: edge
    revision: 4
    resources:
      mimir-image: 2
    scale: 3
    options:
      alertmanager: true
      compactor: true
      distributor: true
      ingester: true
      querier: true
      query-frontend: true
      query-scheduler: true
      ruler: true
      store-gateway: true
    constraints: arch=amd64
    storage:
      data: kubernetes,1,1024M
      recovery-data: kubernetes,1,1024M
    trust: true
  traefik:
    charm: traefik-k8s
    channel: latest/stable
    base: ubuntu@20.04/stable
    scale: 1
    constraints: arch=amd64
    storage:
      configurations: kubernetes,1,1024M
    trust: true
relations:
- - mimir:mimir-cluster
  - mimir-worker:mimir-cluster
- - traefik:traefik-route
  - grafana:ingress
- - traefik:ingress
  - alertmanager:ingress
- - grafana:grafana-source
  - alertmanager:grafana-source
- - grafana:grafana-dashboard
  - alertmanager:grafana-dashboard
- - catalogue:ingress
  - traefik:ingress
- - catalogue:catalogue
  - grafana:catalogue
- - catalogue:catalogue
  - alertmanager:catalogue
- - grafana:grafana-source
  - mimir:grafana-source
- - grafana:grafana-dashboard
  - mimir:grafana-dashboards-provider
```
4. Offered the `s3` endpoint from the `ceph-radosgw` and related it with `mimir`.

### Environment

I'm using:

* Juju 3.3.1
* Microk8s 1.28
* Ceph Quincy (except for the object store which is Reef, because it implements the `s3` interface)

### Relevant log output

```shell
Endpoint url cannot have fully qualified paths.
initialize s3 client
github.com/thanos-io/objstore/providers/s3.NewBucketWithConfig
	/root/parts/mimir/build/vendor/github.com/thanos-io/objstore/providers/s3/s3.go:273
github.com/grafana/mimir/pkg/storage/bucket/s3.NewBucketClient
	/root/parts/mimir/build/pkg/storage/bucket/s3/bucket_client.go:27
github.com/grafana/mimir/pkg/storage/bucket.NewClient
	/root/parts/mimir/build/pkg/storage/bucket/client.go:163
github.com/grafana/mimir/pkg/mimir.(*Mimir).initUsageStats
	/root/parts/mimir/build/pkg/mimir/modules.go:971
github.com/grafana/dskit/modules.(*Manager).initModule
	/root/parts/mimir/build/vendor/github.com/grafana/dskit/modules/modules.go:136
github.com/grafana/dskit/modules.(*Manager).InitModuleServices
	/root/parts/mimir/build/vendor/github.com/grafana/dskit/modules/modules.go:108
github.com/grafana/mimir/pkg/mimir.(*Mimir).Run
	/root/parts/mimir/build/pkg/mimir/mimir.go:812
main.main
	/root/parts/mimir/build/cmd/mimir/main.go:227
runtime.main
	/snap/go/10489/src/runtime/proc.go:267
runtime.goexit
	/snap/go/10489/src/runtime/asm_amd64.s:1650
create usage-stats bucket client
github.com/grafana/mimir/pkg/mimir.(*Mimir).initUsageStats
	/root/parts/mimir/build/pkg/mimir/modules.go:973
github.com/grafana/dskit/modules.(*Manager).initModule
	/root/parts/mimir/build/vendor/github.com/grafana/dskit/modules/modules.go:136
github.com/grafana/dskit/modules.(*Manager).InitModuleServices
	/root/parts/mimir/build/vendor/github.com/grafana/dskit/modules/modules.go:108
github.com/grafana/mimir/pkg/mimir.(*Mimir).Run
	/root/parts/mimir/build/pkg/mimir/mimir.go:812
main.main
	/root/parts/mimir/build/cmd/mimir/main.go:227
runtime.main
	/snap/go/10489/src/runtime/proc.go:267
runtime.goexit
	/snap/go/10489/src/runtime/asm_amd64.s:1650
error initialising module: usage-stats
github.com/grafana/dskit/modules.(*Manager).initModule
	/root/parts/mimir/build/vendor/github.com/grafana/dskit/modules/modules.go:138
github.com/grafana/dskit/modules.(*Manager).InitModuleServices
	/root/parts/mimir/build/vendor/github.com/grafana/dskit/modules/modules.go:108
github.com/grafana/mimir/pkg/mimir.(*Mimir).Run
	/root/parts/mimir/build/pkg/mimir/mimir.go:812
main.main
	/root/parts/mimir/build/cmd/mimir/main.go:227
runtime.main
	/snap/go/10489/src/runtime/proc.go:267
runtime.goexit
	/snap/go/10489/src/runtime/asm_amd64.s:1650
```
```


### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lucabello**
## Issue

Closes #44.

Also does a `charmcraft fetch-lib`.

## Solution

Strip the scheme from the S3 endpoint when building the Mimir configuration.

### CODE REVIEW

File: src/mimir_coordinator.py [hunk 1]
```diff
@@ -259,6 +260,11 @@ def _build_blocks_storage_config(self) -> Dict[str, Any]:
         }
 
     def _build_s3_storage_config(self, s3_config_data: _S3ConfigData) -> Dict[str, Any]:
+        # Remove the scheme from the s3 endpoint for the Mimir config
+        parsed_url = urlparse(s3_config_data.endpoint)
+        scheme = "%s://" % parsed_url.scheme
+        s3_config_data.endpoint = parsed_url.geturl().replace(scheme, "", 1)
+        # Return the s3 configuration
```
Comment (IbraAoad) [inline src/mimir_coordinator.py:267]: Does it make sense to move this to be a pydantic validator of the `endpoint` field in the `_S3ConfigData` class instead?

**Review by IbraAoad — APPROVED** [review]

File: src/mimir_coordinator.py (hunk 1 above)
Comment (lucabello) [reply src/mimir_coordinator.py:267]: The problem is I don't think we have control on what we receive over that relation

File: src/mimir_coordinator.py [hunk 2]
```diff
@@ -259,6 +260,11 @@ def _build_blocks_storage_config(self) -> Dict[str, Any]:
         }
 
     def _build_s3_storage_config(self, s3_config_data: _S3ConfigData) -> Dict[str, Any]:
+        # Remove the scheme from the s3 endpoint for the Mimir config
+        parsed_url = urlparse(s3_config_data.endpoint)
+        scheme = "%s://" % parsed_url.scheme
+        s3_config_data.endpoint = parsed_url.geturl().replace(scheme, "", 1)
```
Comment (sed-i) [inline src/mimir_coordinator.py:266]: Nitpick: `replace` is more than `strip_prefix` (which isn't available on py3.8).
How about:
```python
>>> url = "scheme://localhost:9090"
>>> re.sub(fr'^{urlparse(url).scheme}://', '', url)
'localhost:9090'
```

File: tests/unit/test_config.py [hunk 3]
```diff
@@ -73,7 +73,7 @@ def test_build_config_without_s3_data(self):
 
     def test_build_s3_storage_config(self):
         raw_s3_config_data = {
-            "endpoint": "s3.com:port",
+            "endpoint": "http://s3.com:port",
```
Comment (sed-i) [inline tests/unit/test_config.py:76]: Are we guaranteed to have a scheme in there? Should the stripping code take caution?

**Review by sed-i — APPROVED** [review]

File: src/mimir_coordinator.py (hunk 1 above)
Comment (IbraAoad) [reply src/mimir_coordinator.py:267]: sure but what I mean is 

```
    @validator("endpoint")
    def remove_scheme(cls, v):
        parsed_url = urlparse(v)
        scheme = "%s://" % parsed_url.scheme
        return parsed_url.geturl().replace(scheme, "", 1)
```

so regardless of what's coming in we remove the scheme, also we might need to accommodate for cases which the s3 endpoint will be an internal cluster storage (already has no scheme)

File: tests/unit/test_config.py (hunk 3 above)
Comment (lucabello) [reply tests/unit/test_config.py:76]: The stripping function (before and after your suggestion to use `re.sub()`) returns the base URL regardless; I'll add a separate test for that, just to be safe.

File: src/mimir_coordinator.py (hunk 1 above)
Comment (lucabello) [reply src/mimir_coordinator.py]: Done!

**Review by IbraAoad — APPROVED** [review]

**Pull request merged**